File: Linked List/singly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node *next;
};
struct node *head = NULL;
struct node *current = NULL;

void printList(){
    struct node *p = head;
    while(p!=NULL){
        printf("%d ",p->data);
        p = p->next;
    }
}

void insertAtBegin(int data){
    struct node *newNode = (struct node*)malloc(sizeof(struct node));
    newNode->data = data;
    newNode->next = head;
    head = newNode;
}

void insertAtEnd(int data) {
    struct node *newNode = (struct node*)malloc(sizeof(struct node));
    newNode->data = data;
    newNode->next = NULL;

    if (head == NULL) {
        head = newNode;
        return;
    }

    struct node *current = head;
    while (current->next != NULL) {
        current = current->next;
    }
    
    current->next = newNode;
}

void insertAfterNode(struct node *list, int data) {
    struct node *newNode = (struct node*)malloc(sizeof(struct node));
    newNode->data = data;
    newNode->next = list->next;
    list->next = newNode;
}

void deleteAtBegin() {
    if(head == NULL) {
        return;
    }
    struct node* temp = head;
    head = head->next;
    free(temp);
}

void deleteAtEnd() {
    if(head == NULL) {
        return;
    }

    if(head->next == NULL) {
        free(head);
        head = NULL;
        return;
    }
    struct node* temp = head;
    while(temp->next->next != NULL) {
        temp = temp->next;
    }
    free(temp->next);
    temp->next = NULL;

}

void reverse() {
    struct node *prev = NULL;
    struct node *current = head;
    struct node *next = NULL;

    while (current != NULL) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    head = prev;
}

int search(int value) {
    struct node *current = head;
    while (current != NULL) {
        if (current->data == value) {
            return 1;
        }
        current = current->next;
    }
    return 0;
}

int main() {
    
    insertAtEnd(100);
    printList();
    printf("\n");
    insertAtBegin(1);
    printList();
    printf("\n");
    deleteAtBegin();
    printList();
    printf("\n");
    insertAtBegin(2);
    insertAfterNode(head->next, 50);
    insertAtEnd(3);
    deleteAtEnd();
    printList();
    printf("\n");
   
    insertAtEnd(1);
    insertAtEnd(10);
    printList();
    printf("\n");
    reverse();
    printList();
    
    int valueToSearch = 2;
    if (search(valueToSearch)) {
        printf("\nValue %d found in the list.\n", valueToSearch);
    } else {
        printf("\nValue %d not found in the list.\n", valueToSearch);
    }

    return 0;

}

File: Linked List/singly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None 
    
    def print_list(self):
        current = self.head
        while current:
            print(f"{current.data} -> ", end="")
            current = current.next
        print("None")

    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return

        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
    
    def insert_at_begin(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def delete_at_begin(self):
        if self.head is None:
            return
        self.head = self.head.next

    def delete_at_end(self):
        if self.head is None:
            return
        
        if self.head.next is None:
            self.head = None
            return
        
        current = self.head
        while current.next.next is not None:
            current = current.next
        current.next = None
    
    def reverse(self):
        prev = None
        current = self.head

        while current is not None:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node 
        self.head = prev
    
    def search(self, value):
        current = self.head
        while current is not None:
            if current.data == value:
                return 1
            current = current.next
        return 0

if __name__ == "__main__":
    sll = SinglyLinkedList()

    sll.print_list()
    sll.insert_at_end(1)
    sll.insert_at_end(2)
    sll.insert_at_end(3)
    sll.print_list()
    sll.insert_at_begin(10)
    sll.insert_at_begin(12)
    sll.print_list()
    sll.insert_at_begin(19)
    sll.print_list()
    sll.delete_at_begin()
    sll.print_list()
    sll.insert_at_end(100)
    sll.print_list()
    sll.delete_at_end()
    sll.print_list()
    sll.reverse()
    sll.print_list()

    value = 15
    if sll.search(value):
        print(f"Found {value}")
    else:
        print(f"Not Found")
